File: 460-lfu-cache/460-lfu-cache.cpp

struct Node
{
    Node *prev, *next;
    int val;

public:
    Node()
    {
        prev = NULL;
        next = NULL;
        val = INT_MAX;
    }
    Node(int k)
    {
        prev = NULL;
        next = NULL;
        val = k;
    }
};
struct List
{
    Node *head, *tail;
    int size;

public:
    List()
    {
        head = new Node();
        tail = new Node();
        head->next = tail;
        tail->prev = head;
        size = 0;
    }

    Node *pushFront(int key)
    {
        Node* node = new Node(key);
        node->prev = head;
        node->next = head->next;
        node->next->prev = node;
        head->next = node;
        size++;
        return node;
    }

    int getSize()
    {
        return size;
    }

    Node *popBack()
    {
        if (head->next != tail)
        {
            
            auto toBeDeleted = tail->prev;
            toBeDeleted->prev->next = toBeDeleted->next;
            toBeDeleted->next->prev = toBeDeleted->prev;
            size--;
            return toBeDeleted;
        }
        else
        {
            // cout<<"not deleting "<<endl;
            return NULL;
        }
    }

    void remove(Node *node)
    {
        if (node->prev && node->next)
        {

            auto beforeDelete = node->prev;
            auto afterDelete = node->next;
            beforeDelete->next = afterDelete;
            afterDelete->prev = beforeDelete;
            size--;
        }
    }
};

class LFUCache
{
public:
    int maxSize, curSize, minFreq;
    map<int, int> keyFreq, keyVal;
    map<int, List *> freqList;
    map<int, Node *> keyNode;

    LFUCache(int capacity)
    {
        maxSize = capacity;
        minFreq = 0;
        curSize = 0;
    }

    void put(int key, int value)
    {
        if (maxSize == 0)
            return;

        if (keyFreq.count(key))
        {
            keyVal[key] = value;
            updateFreqList(key);
        }
        else
        {
            if (curSize >= maxSize)
            {
                
                List* l = freqList[minFreq];
                // cout<<"l's ("<<minFreq<<") first element is: "<<l->head->next->val<<endl;
                auto removedNode = l->popBack();
                keyVal.erase(removedNode->val);
                keyFreq.erase(removedNode->val);
                keyNode.erase(removedNode->val);
                curSize--;
            }
            minFreq = 1;
            curSize++;
            keyFreq[key] = 1;
            keyVal[key] = value;
            if (!freqList.count(minFreq))
                freqList[minFreq] = new List();
            keyNode[key] = freqList[minFreq]->pushFront(key);
            // auto temp = freqList[minFreq]->head;
            // while(temp){
            //     temp = temp->next;
            // }
            // cout<<endl;
        }
        // cout<<key<<endl;
//         for(auto m:keyFreq){
//             cout<<m.first<<" "<<m.second<<"\t";
//         }
        
        // cout<<endl<<"minFreq: "<<minFreq<<endl;
    }

    void updateFreqList(int key)
    {
        auto freq = keyFreq[key];
        auto newFreq = freq + 1;

        // remove from previous list
        // for(auto m:keyFreq){
        //     cout<<m.first<<" "<<m.second<<"\t";
        // }
        freqList[freq]->remove(keyNode[key]);
        // cout<<"updating freq list, freq:"<<freq<<" minFreq:"<<minFreq<<endl;
        if (freqList[freq]->getSize() == 0 && minFreq == freq)
        {
            minFreq++;
        }
        keyNode.erase(key);
        // keyFreq.erase(key);
        if (!freqList.count(newFreq))
        {
            freqList[newFreq] = new List();
        }
        auto node = freqList[newFreq]->pushFront(key);
        keyFreq[key] = newFreq;
        keyNode[key] = node;
    }

    int get(int key)
    {
        if (!keyVal.count(key))
        {
            return -1;
        }
        else
        {
            auto val = keyVal[key];
            updateFreqList(key);
            return val;
        }
    }
};

File: 460-lfu-cache/460-lfu-cache.py
class Node:
    def __init__(self,key=inf,val=inf,prev=None,next=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next
        self.freq = 0
    def __str__(self):
        return "(key="+str(self.key)+", val="+str(self.val)+")"
class DLL:
    def __init__(self):
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.sz = 0
        
    def remove(self,node):
        p,n = node.prev,node.next
        p.next = n
        n.prev = p
        node.prev = None
        node.next = None
        self.sz-=1
    def removeLast(self):
        if self.sz!=0:
            lastNode = self.tail.prev
            secondLast = lastNode.prev
            secondLast.next = self.tail
            self.tail.prev = secondLast
            lastNode.prev = None
            lastNode.next = None
            self.sz -= 1
            return lastNode
    
    def addFirst(self,node):
        oldFirstNode = self.head.next
        node.prev = self.head
        node.next = oldFirstNode
        oldFirstNode.prev = node
        self.head.next = node
        self.sz+=1            
            
    def __len__(self):
        return self.sz
    def __str__(self):
        ans = ""
        temp = self.head.next
        while temp!=self.tail:
            ans+= str(temp)+" "
            temp = temp.next
        return ans
class LFUCache:
    
    def show(self):
        print("minFreq: ",self.minFreq)
        print("DLL View")
        for k in self.freqDLL:
            if len(self.freqDLL[k])>0:print("freq:",k, "DLL:",self.freqDLL[k])
        print("DLL View end")
        print("freqNode view")
        for f in self.keyNode:
            print(f,self.keyNode[f])
        print("end freqNode view")
        
    def __init__(self, capacity: int):
        self.freqDLL = defaultdict(DLL)
        self.keyNode = {}
        self.cap = capacity
        self.usedCap = 0
        self.minFreq = 1
        
    def get(self, key: int) -> int:
        # print("op: get",key)
        # self.show()
        if key in self.keyNode:
            node = self.keyNode[key]
            freq = node.freq
            val = node.val
            self.freqDLL[freq].remove(node)
            if len(self.freqDLL[freq])==0 and freq==self.minFreq:
                self.minFreq+=1
            node.freq+=1
            self.freqDLL[freq+1].addFirst(node)
            return val
        else: return -1

    def put(self, key: int, value: int) -> None:
        if self.cap<=0: return
        # print("op: put",key,value)
        # self.show()
        if key in self.keyNode:
            node = self.keyNode[key]
            self.freqDLL[node.freq].remove(node)
            if len(self.freqDLL[node.freq])==0 and node.freq==self.minFreq:
                self.minFreq+=1
        else:
            self.usedCap+=1
            if self.usedCap>self.cap:
                self.usedCap -= 1
                removedNode = self.freqDLL[self.minFreq].removeLast()
                del self.keyNode[removedNode.key]
                if len(self.freqDLL[self.minFreq])==0:
                    self.minFreq += 1
            node = Node(key=key,val=value)
            
        node.freq += 1
        node.val = value
        self.minFreq = min(self.minFreq,node.freq)
        self.freqDLL[node.freq].addFirst(node)
        self.keyNode[key]=node
        
